### Homework 3 Numpy Arrays and Pandas

1. Open `cars.csv` in jupyter and look at the data. (Double click `cars.csv` to open in a new tab).
2. We are converting each **column** into a numpy array.
3. Notice we have text, integer and float data. 
4. Take care not to overwrite your starter code.

In [2]:
#Starter code. Do NOT modify this cell
import numpy as np
import pandas as pd

# Read in dataset
df = pd.read_csv('cars.csv',delimiter=';')

# Convert each column to an np.array
car = df.Car.to_numpy()
mpg = df.MPG.to_numpy()
cylinders = df.Cylinders.to_numpy()
displacement = df.Displacement.to_numpy()
horsepower = df.Horsepower.to_numpy()
weight = df.Weight.to_numpy()
acceleration = df.Acceleration.to_numpy()
model = df.Model.to_numpy()
origin = df.Origin.to_numpy()

# create an array of headers
headers = np.array(['Car', 'MPG', 'Cylinders', 'Displacement', 'Horsepower', 'Weight','Acceleration', 'Model', 'Origin'])


1. (2 pts) What are the numpy datatypes for the `car`,`mpg`, `horsepower` and `origin` arrays? (Hint: Not `type(car)` and use `print()` to print the numpy type.)

In [40]:
print(df["Car"].dtype)
#Car is a string, or sentence datatype

str


In [41]:
print(df["MPG"].dtype)
#MPG is a float, or decimal datatype

float64


In [42]:
print(df["Horsepower"].dtype)
#Horsepower is a float, or decimal datatype

float64


In [43]:
print(df["Origin"].dtype)
#Origin is a string, or sentence datatype

str


2. (1 pt) `numpy` is often imported as `np`. What is this action called? Why might you do this? 

In [44]:
"""
this is called giving it an alias
we do this for a few reasons
first is that it makes the name shorter and easier/quicker to call
second is that another module might have the same or a similar name and it avoids confusion
"""

'\nthis is called giving it an alias\nwe do this for a few reasons\nfirst is that it makes the name shorter and easier/quicker to call\nsecond is that another module might have the same or a similar name and it avoids confusion\n'

3. (2 pts) Each of the arrays (cars, mpg, cylinders, etc..) has how many dimensions? Show how to find this using code. 

In [8]:
df['Car'].ndim

1

In [9]:
df['MPG'].ndim

1

In [13]:
df['Cylinders'].ndim

1

In [14]:
df['Displacement'].ndim

1

In [15]:
df['Horsepower'].ndim

1

In [16]:
df['Weight'].ndim

1

In [17]:
df['Acceleration'].ndim

1

In [18]:
df['Model'].ndim

1

In [19]:
df['Origin'].ndim

1

Each array has one dimension. 

4. (3 pts) What is the ___ of `horsepower`?
- sum
- min
- max
- median
- mode

In [20]:
print(np.sum(df['Horsepower']))

42033.0


In [21]:
print(np.min(df['Horsepower']))

0.0


In [22]:
print(np.max(df['Horsepower']))

230.0


In [23]:
print(np.median(df['Horsepower']))

93.5


In [28]:
from scipy import stats
mode = stats.mode(df['Horsepower'])
print(mode.mode)

150.0


The sum is 42033, the min is 0, the max is 230, the median is 93.5, and the mode is 150. 

5. (2 pts) Select the last element in each subarray (do this with one operation.)

In [106]:
select_more = np.array([[3,2,1],
                      [4,5,6],
                      [7,8,9]])
# returns array([1, 6, 9])
array1 = select_more[-3]
array2 = select_more[-2]
array3 = select_more[-1]
last_elements = np.array([array1[-1], array2[-1], array3[-1]])
last_elements

array([1, 6, 9])

6. (2 pts) a) what is boolean indexing?
<br/> b) Use boolean indexing to create an array of `True/False` using `origin` and the value 'US'.

In [58]:
#Boolean indexing turns values to a true/false value based on a specific criteria
origin_bool = df["Origin"] == "US"
print(origin_bool)


0       True
1       True
2       True
3       True
4       True
       ...  
401     True
402    False
403     True
404     True
405     True
Name: Origin, Length: 406, dtype: bool


7. (2 pts) Apply your index (from above) to the `car` array. You should now only see cars that are of US origin. 

In [62]:
from sklearn.preprocessing import LabelEncoder

class_labels = LabelEncoder()
df['US'] = class_labels.fit_transform(origin_bool)

df[df['US'] == 1]

,Car,MPG,Cylinders,Displacement,Horsepower,Weight,Acceleration,Model,Origin,US
0,Chevrolet Chevelle Malibu,18.0,8,307.0,130.0,3504.0,12.0,70,US,1
1,Buick Skylark 320,15.0,8,350.0,165.0,3693.0,11.5,70,US,1
2,Plymouth Satellite,18.0,8,318.0,150.0,3436.0,11.0,70,US,1
3,AMC Rebel SST,16.0,8,304.0,150.0,3433.0,12.0,70,US,1
4,Ford Torino,17.0,8,302.0,140.0,3449.0,10.5,70,US,1
...,...,...,...,...,...,...,...,...,...,...
400,Chevrolet Camaro,27.0,4,151.0,90.0,2950.0,17.3,82,US,1
401,Ford Mustang GL,27.0,4,140.0,86.0,2790.0,15.6,82,US,1
403,Dodge Rampage,32.0,4,135.0,84.0,2295.0,11.6,82,US,1
404,Ford Ranger,28.0,4,120.0,79.0,2625.0,18.6,82,US,1


### Use numpy or pandas to answer the following questions.

8. (3 pts) What is the average horsepower for `Europe` cars? (Hint: create a boolean index of 'US' apply it to horsepower and perform the calculation.)

In [78]:
europe_bool = df["Origin"] == "Europe"
class_labels = LabelEncoder()
df['Europe'] = class_labels.fit_transform(europe_bool)

europe_df = df[df['Europe'] == 1]
europe_df

,Car,MPG,Cylinders,Displacement,Horsepower,Weight,Acceleration,Model,Origin,US,Europe
10,Citroen DS-21 Pallas,0.0,4,133.0,115.0,3090.0,17.5,70,Europe,0,1
25,Volkswagen 1131 Deluxe Sedan,26.0,4,97.0,46.0,1835.0,20.5,70,Europe,0,1
26,Peugeot 504,25.0,4,110.0,87.0,2672.0,17.5,70,Europe,0,1
27,Audi 100 LS,24.0,4,107.0,90.0,2430.0,14.5,70,Europe,0,1
28,Saab 99e,25.0,4,104.0,95.0,2375.0,17.5,70,Europe,0,1
...,...,...,...,...,...,...,...,...,...,...,...
366,Peugeot 505s Turbo Diesel,28.1,4,141.0,80.0,3230.0,20.4,81,Europe,0,1
367,Saab 900s,0.0,4,121.0,110.0,2800.0,15.4,81,Europe,0,1
368,Volvo Diesel,30.7,6,145.0,76.0,3160.0,19.6,81,Europe,0,1
383,Volkswagen Rabbit l,36.0,4,105.0,74.0,1980.0,15.3,82,Europe,0,1


In [81]:
print(np.average(europe_df["Horsepower"]))
#the average horsepower for european cars is about 78.781

78.78082191780823


9. (3 pts) On average are Japanese cars more fuel-efficient than US cars? Write a function that can handle this comparison.  

In [88]:
"""
creates a column for japanese cars, then finds the average MPG
of japanese cars, then finds the average MPG of US cars. Returns
the two averages in a sentence 
"""
japan_bool = df["Origin"] == "Japan"
class_labels = LabelEncoder()
df['Japan'] = class_labels.fit_transform(japan_bool)
df

,Car,MPG,Cylinders,Displacement,Horsepower,Weight,Acceleration,Model,Origin,US,Europe,Japan
0,Chevrolet Chevelle Malibu,18.0,8,307.0,130.0,3504.0,12.0,70,US,1,0,0
1,Buick Skylark 320,15.0,8,350.0,165.0,3693.0,11.5,70,US,1,0,0
2,Plymouth Satellite,18.0,8,318.0,150.0,3436.0,11.0,70,US,1,0,0
3,AMC Rebel SST,16.0,8,304.0,150.0,3433.0,12.0,70,US,1,0,0
4,Ford Torino,17.0,8,302.0,140.0,3449.0,10.5,70,US,1,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...
401,Ford Mustang GL,27.0,4,140.0,86.0,2790.0,15.6,82,US,1,0,0
402,Volkswagen Pickup,44.0,4,97.0,52.0,2130.0,24.6,82,Europe,0,1,0
403,Dodge Rampage,32.0,4,135.0,84.0,2295.0,11.6,82,US,1,0,0
404,Ford Ranger,28.0,4,120.0,79.0,2625.0,18.6,82,US,1,0,0


In [ ]:
def japan_vs_us()

10. (3 pts) What is the correlation coefficient of `mpg` and `weight`? What is a correlation coefficient? What does the result indicate about the relationship between mpg and weight? 

In [93]:
np.corrcoef(mpg, weight)

array([[ 1.        , -0.78750994],
       [-0.78750994,  1.        ]])

The correlation coefficient of mpg and weight is -0.7875. 
The correlation coefficient is a value between -1 and 1
that indicates how closely correlated the two variables 
are. The closer to zero the value, the less correlated 
the two variables are. The closer to -1 or 1 the 
value, the more correlated the two variables are. 
The correlation coefficient of mpg and weight 
is pretty high, meaning that the variables are pretty closely
correlated. The negative means that they are negatively
correlated. This means that as mpg goes up, weight goes 
down, and vice versa. 

11. (2 pts) Create a new pandas dataframe that has the average mpg, weight, and horsepower per country of origin

In [ ]:
averages_df = pd.DataFrame(data = 

Extra credit (2 pts): Write a function to find the deepest level of nested brackets
<br/>Sample Input: {"a":1,"b":2,"c":{"d":3,"e":4,"f":{"g":5,"h":6}}}
<br/>Sample Output: There are 3 nested layers

In [107]:
def nested_brackets(input):
    a = 0
    for x in (input):
        if x == "{":
            a += 1
        else: 
            a += 0
    return (f'There are {a} nested brackets')

nested_brackets('{"a":1,"b":2,"c":{"d":3,"e":4,"f":{"g":5,"h":6}}}')



'There are 3 nested brackets'